pkg/clients/postgresql/postgresql.go

```diff
@@ -3,38 +3,76 @@ package postgresql
 import (
 	"context"
 	"database/sql"
+	"fmt"
 
 	"github.com/crossplane-contrib/provider-sql/pkg/clients/xsql"
 
 	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
 	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
 )
 
+const (
+	panicNilPtr = "function call on nil pointer"
+)
+
 type postgresDB struct {
 	dsn      string
+	username string
+	password string
+	database string
 	endpoint string
 	port     string
 }
 
 // New returns a new PostgreSQL database client.
+// The default database name is an empty string.
+// The underlying pq library will default to either
+// using the value of PGDATABASE, or if unset, the
+// hardcoded db name 'postgres'.
+// The db name can be overridden later using SetDatabase,
+// which will trigger a re-format of the DSN.
 func New(creds map[string][]byte) xsql.DB {
 	// TODO(negz): Support alternative connection secret formats?
-	endpoint := string(creds[xpv1.ResourceCredentialsSecretEndpointKey])
-	port := string(creds[xpv1.ResourceCredentialsSecretPortKey])
-	return postgresDB{
-		dsn: "postgres://" +
-			string(creds[xpv1.ResourceCredentialsSecretUserKey]) + ":" +
-			string(creds[xpv1.ResourceCredentialsSecretPasswordKey]) + "@" +
-			endpoint + ":" +
-			port,
-		endpoint: endpoint,
-		port:     port,
+	db := postgresDB{
+		username: string(creds[xpv1.ResourceCredentialsSecretUserKey]),
+		password: string(creds[xpv1.ResourceCredentialsSecretPasswordKey]),
+		endpoint: string(creds[xpv1.ResourceCredentialsSecretEndpointKey]),
+		port:     string(creds[xpv1.ResourceCredentialsSecretPortKey]),
 	}
+	db.formatDSN()
+	return &db
+}
+
+// formatDSN sets the DSN that this client will use
+// to connect to PostgreSQL.
+func (c *postgresDB) formatDSN() {
+	if c == nil {
+		panic(panicNilPtr)
```

Comment:
Does this really add any benefit over just letting it panic with a nil dereference later on?

Reply:
Nope, force of habit - better to panic at the point of dereference as you said 👍